server/src/main/java/org/elasticsearch/index/codec/tsdb/es819/ES819TSDBDocValuesProducer.java

```diff
@@ -438,8 +445,25 @@ BlockLoader.Block tryReadAHead(BlockLoader.BlockFactory factory, BlockLoader.Doc
                         BytesRef b = lookupOrd(Math.toIntExact(startValue));
                         return factory.constantBytes(BytesRef.deepCopyOf(b), docCount - offset);
                     }
-                    // TODO: Since ordinals are sorted, start at 0 (offset by startValue), scan until lastValue,
-                    // then fill remaining positions with lastValue.
+                    var ordinalReader = denseOrds.sortedOrdinalReader();
+                    if (ordinalReader != null) {
+                        try (var builder = factory.singletonOrdinalsBuilder(this, docCount - offset, true)) {
+                            int docIndex = offset;
+                            while (docIndex < docCount) {
+                                long ord = ordinalReader.readValueAndAdvance(docs.get(docIndex));
```

Comment:
nit: can we immediately convert this to an int here? I think we would then need to do same for `lastValue` and `startValue`, but that should be given that we only care about ints?